spopt/locate/coverage.py

```diff
@@ -0,0 +1,395 @@
+
+        mclp.aij = np.zeros(cost_matrix.shape)
+        mclp.aij[cost_matrix <= max_coverage] = 1
+        weights = np.reshape(weights, (cost_matrix.shape[0], 1))
+
+        mclp.__add_obj(weights, r_cli)
+        FacilityModelBuilder.add_maximal_coverage_constraint(
+            mclp, mclp.problem, mclp.aij, r_fac, r_cli
+        )
+        FacilityModelBuilder.add_facility_constraint(mclp, mclp.problem, p_facilities)
+
+        return mclp
+
+    @classmethod
+    def from_geodataframe(
+        cls,
+        gdf_demand: GeoDataFrame,
+        gdf_fac: GeoDataFrame,
+        demand_col: str,
+        facility_col: str,
+        weights_cols: str,
+        max_coverage: float,
+        p_facilities: int,
+        distance_metric: str = "euclidean",
+        name: str = "MCLP",
+    ):
+        """
+        Create a MCLP object based on geodataframes. Calculate the cost matrix between demand and facility,
+        and then use from_cost_matrix method.
+
+        Parameters
+        ----------
+        gdf_demand: geopandas.GeoDataFrame
+            demand geodataframe with point geometry
+        gdf_fac: geopandas.GeoDataframe
+            facility geodataframe with point geometry
+        demand_col: str
+            demand geometry column name
+        facility_col: str
+            facility candidate sites geometry column name
+        weights_cols: str
+            weight column name representing service load or demand
+        max_coverage: float
+            maximum acceptable service distance by problem
+        p_facilities: int
+            number of facilities to be located
+        distance_metric: str, default="euclidean"
+            metrics supported by :method: `scipy.spatial.distance.cdist` used for the distance calculations
+        name: str, default="MCLP"
+            name of the problem
+
+        Returns
+        -------
+        MCLP object
+        """
+        service_load = gdf_demand[weights_cols].to_numpy()
+        dem = gdf_demand[demand_col]
+        fac = gdf_fac[facility_col]
+
+        dem_type_geom = dem.geom_type.unique()
+        fac_type_geom = fac.geom_type.unique()
+
+        if len(dem_type_geom) > 1 or not "Point" in dem_type_geom:
+            warnings.warn(
+                "Demand geodataframe contains mixed type geometries or is not a point. Be sure deriving centroid from geometries doesn't affect the results.",
+                Warning,
+            )
+            dem = dem.centroid
+
+        if len(fac_type_geom) > 1 or not "Point" in fac_type_geom:
+            warnings.warn(
+                "Facility geodataframe contains mixed type geometries or is not a point. Be sure deriving centroid from geometries doesn't affect the results.",
+                Warning,
+            )
+            fac = fac.centroid
+
+        dem_data = np.array([dem.x.to_numpy(), dem.y.to_numpy()]).T
+        fac_data = np.array([fac.x.to_numpy(), fac.y.to_numpy()]).T
+
+        distances = np.array([])
```

Comment:
Does `distances` need to be declared here and again in line 350?

* also check this in MCLP, *p*-median, and *p*-center